Plugins: Fix plugins CDN not working when plugins are not in domain's root path

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Plugins CDN path construction is broken if the plugins CDN root is in a "subfolder", ~~and if the path does not contain `plugin-cdn`.~~

This is what we use in production, and some files are not loading properly:

`https://grafana-assets.grafana.net/plugins/{id}/{version}/public/plugins/{id}/{assetPath}`

**Why do we need this feature?**

Fixes a bug in plugins CDN assets path construction.

**Who is this feature for?**

Hosted Grafana users

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

**Special notes for your reviewer**:

xnyo committed:
Plugins CDN: Add support for different CDN root path

xnyo committed:
Plugins CDN: Make frontendsettings return the correct CDN base path

xnyo committed:
Update comments

xnyo committed:
Fix version detection

xnyo committed:
Undo frontend changes

xnyo committed:
Fix system.js asset path construction

jackw committed:
fix(plugins): translate all plugin css asset paths loaded via cdn

jackw committed:
refactor(plugins): rename extractPluginNameVersionFromUrl and add comments

xnyo committed:
Fix typo in comment

Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

mckn reviewed (approved):
LGTM!

wbrowne commented on pkg/plugins/pluginscdn/pluginscdn.go:
```diff
@@ -3,15 +3,14 @@ package pluginscdn
-	// systemJSCDNKeyword is the path prefix used by system.js to identify the plugins CDN.
-	systemJSCDNKeyword = "plugin-cdn"
+	// systemJSCDNURLTemplate is a special path template used by system.js to identify plugin CDN assets
+	systemJSCDNURLTemplate = "plugin-cdn/{id}/{version}/public/plugins/{id}/{assetPath}"
```
Considering this is baked into the code this way, should the plugin CDN URL config value maybe instead be just the domain instead? 🤔

xnyo replied:
Yeah I think it makes more sense and it'll be consistent. I'll change that! :+1:

xnyo committed:
Hardcode CDN URL structure

/{id}/{version}/public/plugins/{id}/{assetPath} is not required anymore in the cdn url template config

xnyo commented on pkg/plugins/pluginscdn/pluginscdn.go:
```diff
@@ -30,7 +31,7 @@ func ProvideService(cfg *config.Cfg) *Service {
 func (s *Service) NewCDNURLConstructor(pluginID, pluginVersion string) URLConstructor {
 	return URLConstructor{
-		cdnURLTemplate: s.cfg.PluginsCDNURLTemplate,
+		cdnURLTemplate: strings.TrimRight(s.cfg.PluginsCDNURLTemplate, "/") + "/" + cdnAssetPathTemplate,
```
Config change:

old:
```
[plugins]
cdn_base_url = http://127.0.0.1:1234/plugins/{id}/{version}/public/plugins/{id}/{assetPath}
```

new:
```
[plugins]
cdn_base_url = http://127.0.0.1:1234/plugins
```

wbrowne reviewed (approved):
🚀
